### PR TITLE
Add signup functionality

### PR DESCRIPTION
**Pull request by danifosy**
- saves user name and user type in local storage
- use reducer to handle login/logout
- useContext for global state
- protects private routes

sources: 
- https://www.freecodecamp.org/news/state-management-with-react-hooks/
- https://fettblog.eu/typescript-react/events/
- https://reactrouter.com/web/example/auth-workflow
- https://stackoverflow.com/questions/56322667/how-to-type-a-form-component-with-onsubmit

Note: I obviously did not do this by myself but with lots of help

### CODE REVIEW

File: src/app/App.tsx [hunk 1]
```diff
@@ -6,42 +6,125 @@ import ContactForm from './pages/ContactForm/ContactForm';
...
 import SignUpPage from './pages/SignUpPage/SignUpPage';
 import LoginPage from './pages/LoginPage/LoginPage';
 import MainPage from './pages/MainPage/MainPage';
 import LandingPage from './pages/LandingPage/LandingPage';
 import FavoritesPage from './pages/FavoritesPage/FavoritesPage';
 import MainPageARA from './pages/MainPageARA/MainPageARA';
 import UploadPage from './pages/UploadPage/UploadPage';
+import PrivateRoute from './components/PrivateRoute/PrivateRoute';
+
+export type UserType = 'ARA' | 'user';
```
Comment (marwinburesch) [inline src/app/App.tsx:18]: what's ARA

File: src/app/App.tsx [hunk 2]
```diff
@@ -6,42 +6,125 @@ import ContactForm from './pages/ContactForm/ContactForm';
...
+      return {
+        ...state,
+        isAuthenticated: false,
+        userName: undefined,
+        userType: undefined,
+      };
+    default:
+      return state;
+  }
+};
```
Comment (marwinburesch) [inline src/app/App.tsx:97]: Good stuff, but you should move this to separate files. Also, you should consider separating action and payload

File: src/app/components/PrivateRoute/PrivateRoute.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import { Redirect, Route } from 'react-router-dom';
+import { AuthContext } from '../../App';
+import type { ReactNode } from 'react';
+
+export default function PrivateRoute({
+  children,
+  ...rest
```
Comment (marwinburesch) [inline src/app/components/PrivateRoute/PrivateRoute.tsx:8]: You could rename this to `props`

**Review by marwinburesch — COMMENTED** [review]
good stuff
I am unsure about `SignUpPage.tsx` the interface seems weird, but I don't have a good suggestion...

File: src/app/App.tsx (hunk 1 above)
Comment (danifosy) [reply src/app/App.tsx:18]: Animal Right's Activist

File: src/app/App.tsx (hunk 2 above)
Comment (danifosy) [reply src/app/App.tsx:97]: Yeah I thought so too, might refactor when there's time, I'm gonna put your suggestion into an issue.

**Pull request merged**